source/geod24/concurrency.d

```diff
@@ -1603,3 +1604,449 @@ private:
+        {
+            T msg;
+            ChannelContext!T new_context;
+            new_context.msg_ptr = &msg;
+            new_context.mutex = mutex;
+            new_context.condition = condition;
+            this.recvq.insertBack(new_context);
+            this.mutex.unlock();
+            new_context.wait();
+            return msg;
+        }
+
+        auto scheduler = thisScheduler;
+        if (Fiber.getThis() && scheduler !is null)
+        {
+            return waitForSender(null, scheduler.newCondition(null));
+        }
+        else
+        {
+            auto mutex = new Mutex();
+            return waitForSender(mutex, new Condition(mutex));
+        }
+    }
+
+
+    /***************************************************************************
+
+	    Return the received message.
+
+        Params:
+            msg = point of message to send
+
+        Return:
+            If successful, return true, otherwise return false
+
+    ***************************************************************************/
+
+    public bool tryReceive (T* msg)
+    {
+        assert(msg !is null);
+
+        this.mutex.lock();
+
+        if (this.closed)
+        {
+            this.mutex.unlock();
+            return false;
+        }
+
+        if (!this.sendq.empty)
+        {
+            ChannelContext!T context = this.sendq.front;
+            this.sendq.removeFront();
+            *(msg) = context.msg;
+            this.mutex.unlock();
+            context.notify();
+            return true;
+        }
+
+        if (!this.queue.empty)
+        {
+            *(msg) = this.queue.front;
+            this.queue.removeFront();
+            assert(this.mcount > 0, "mcount was zero!");
+            this.mcount--;
+            this.mutex.unlock();
+            return true;
+        }
+
+        this.mutex.unlock();
+        return false;
+    }
+
+
+    /***************************************************************************
+
+        Return closing status
+
+        Return:
+            If channel is closed, return true, otherwise return false
+
+    ***************************************************************************/
+
+    public @property bool isClosed ()
+    {
+        synchronized (this.mutex)
+        {
+            return this.closed;
+        }
+    }
+
+
+    /***************************************************************************
+
+        Called when the channel is no longer in use.
+        This function turns off Fiber and Thread, which were waiting for
+        the message to pass.
+        And remove the message that was in the queue.
+        And once closed, the channel is no longer available for message passing.
+
+    ***************************************************************************/
+
+    public void close ()
+    {
+        ChannelContext!T context;
+
+        this.mutex.lock();
+        scope (exit) this.mutex.unlock();
+
+        this.closed = true;
```

Comment:
Does it make sense to call `close` on a channel multiple times? Because if not then there should be another assert here:

```suggestion
        assert(!this.closed, "Channel was already closed!");
        this.closed = true;
```